### PR TITLE
Region provider

### PR DESCRIPTION
**Pull request by sdelamo**
Cherry picks PR https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/1034 to target 4.3.x branch

Close: https://github.com/micronaut-projects/micronaut-oracle-cloud/pull/1034

### CODE REVIEW

File: oraclecloud-sdk-processor/src/main/java/io/micronaut/oraclecloud/clients/processor/OracleCloudSdkProcessor.java [hunk 1]
```diff
@@ -92,6 +92,11 @@ public class OracleCloudSdkProcessor extends AbstractProcessor {
      * Processing environment option used to specify the client classes to process.
      */
     public static final String OCI_SDK_CLIENT_CLASSES_OPTION = "ociSdkClientClasses";
+
+    private static final String RETURN_BUILDER_STATEMENT_WITH_REGION = "return regionProvider.getRegion() != null ? clientBuilder.region(regionProvider.getRegion()).build(authenticationDetailsProvider) : clientBuilder.build(authenticationDetailsProvider)";
```
Comment (andriy-dmytruk) [inline oraclecloud-sdk-processor/src/main/java/io/micronaut/oraclecloud/clients/processor/OracleCloudSdkProcessor.java:96]: This is not optimal, I think. The region would overwrite an endpoint if set by user, see implementation:

```
protected void setRegion(Region region) {
        this.region = region;
        ...

        Optional<String> endpoint = region.getEndpoint(this.service);
        if (endpoint.isPresent()) {
            this.setEndpoint((String)endpoint.get());
        } else {
            throw new IllegalArgumentException("Endpoint for " + this.service + " is not known in region " + region);
        }
    }
```

Comment (andriy-dmytruk) [reply oraclecloud-sdk-processor/src/main/java/io/micronaut/oraclecloud/clients/processor/OracleCloudSdkProcessor.java:96]: Instead, the region should be injected in the constructor of `AbstractSdkClientFactory` and set in the builder there.

Then user can use `BeanCreatedEventListener<ClientBuilder>` and override region or endpoint explicitly.

Comment (andriy-dmytruk) [reply oraclecloud-sdk-processor/src/main/java/io/micronaut/oraclecloud/clients/processor/OracleCloudSdkProcessor.java:96]: Actually, why do we need this? In most cases the region is set from the `AuthenticationDetailsProvider`, e.g. `ResourcePrincipalAuthenticationDetailsProvider` would set it in the endpoint:
```java
public class ResourcePrincipalAuthenticationDetailsProvider extends AbstractRequestingAuthenticationDetailsProvider implements RegionProvider, ...
```

**Comment by alvarosanchez** [comment]
This includes a breaking change in `AbstractSdkClientFactory`

**Comment by alvarosanchez** [comment]
I see now that `AbstractSdkClientFactory` is `@Internal`, so it's probably ok.

**Pull request merged**